SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp

```diff
@@ -9,69 +9,77 @@
 
 DEFINE_LOG_CATEGORY(LogSpatialNetBitReader);
 
+static thread_local FSpatialNetBitReader* GCurrentReader = nullptr;
+
 FSpatialNetBitReader::FSpatialNetBitReader(USpatialPackageMapClient* InPackageMap, uint8* Source, int64 CountBits, TSet<FUnrealObjectRef>& InDynamicRefs, TSet<FUnrealObjectRef>& InUnresolvedRefs)
 	: FNetBitReader(InPackageMap, Source, CountBits)
 	, DynamicRefs(InDynamicRefs)
-	, UnresolvedRefs(InUnresolvedRefs) {}
+	, UnresolvedRefs(InUnresolvedRefs)
+{
+	// Limitation of using a global TLS pointer, you can have at most a single instance of this object per thread.
+	// There should be no need to have more than one at a given time, but should that be the case we could move to a set per thread instead of a pointer.
+	check(GCurrentReader == nullptr);
+	GCurrentReader = this;
+}
 
-void FSpatialNetBitReader::DeserializeObjectRef(FUnrealObjectRef& ObjectRef)
+FSpatialNetBitReader::~FSpatialNetBitReader()
+{
+	GCurrentReader = nullptr;
+}
+
+void FSpatialNetBitReader::DeserializeObjectRef(FArchive& Archive, FUnrealObjectRef& ObjectRef)
 {
 	int64 EntityId;
-	*this << EntityId;
+	Archive << EntityId;
 	ObjectRef.Entity = EntityId;
-	*this << ObjectRef.Offset;
+	Archive << ObjectRef.Offset;
 
 	uint8 HasPath;
-	SerializeBits(&HasPath, 1);
+	Archive.SerializeBits(&HasPath, 1);
 	if (HasPath)
 	{
 		FString Path;
-		*this << Path;
+		Archive << Path;
 
 		ObjectRef.Path = Path;
 	}
 
 	uint8 HasOuter;
-	SerializeBits(&HasOuter, 1);
+	Archive.SerializeBits(&HasOuter, 1);
 	if (HasOuter)
 	{
 		ObjectRef.Outer = FUnrealObjectRef();
-		DeserializeObjectRef(*ObjectRef.Outer);
+		DeserializeObjectRef(Archive, *ObjectRef.Outer);
 	}
 
-	SerializeBits(&ObjectRef.bNoLoadOnClient, 1);
-	SerializeBits(&ObjectRef.bUseClassPathToLoadObject, 1);
+	Archive.SerializeBits(&ObjectRef.bNoLoadOnClient, 1);
+	Archive.SerializeBits(&ObjectRef.bUseClassPathToLoadObject, 1);
 }
 
-UObject* FSpatialNetBitReader::ReadObject(bool& bUnresolved)
+UObject* FSpatialNetBitReader::ReadObject(FArchive& Archive, USpatialPackageMapClient* PackageMap, bool& bUnresolved)
 {
 	FUnrealObjectRef ObjectRef;
 
-	DeserializeObjectRef(ObjectRef);
+	DeserializeObjectRef(Archive, ObjectRef);
 	check(ObjectRef != FUnrealObjectRef::UNRESOLVED_OBJECT_REF);
 
-	UObject* Value = FUnrealObjectRef::ToObjectPtr(ObjectRef, Cast<USpatialPackageMapClient>(PackageMap), bUnresolved);
+	UObject* Value = FUnrealObjectRef::ToObjectPtr(ObjectRef, PackageMap, bUnresolved);
 
-	if (bUnresolved)
+	if (GCurrentReader != nullptr)
 	{
-		UnresolvedRefs.Add(ObjectRef);
-	}
-	else if (Value && !Value->IsFullNameStableForNetworking())
-	{
-		DynamicRefs.Add(ObjectRef);
+		if (bUnresolved)
+		{
+			GCurrentReader->UnresolvedRefs.Add(ObjectRef);
+		}
+		else if (Value && !Value->IsFullNameStableForNetworking())
+		{
+			GCurrentReader->DynamicRefs.Add(ObjectRef);
+		}
 	}
 
 	return Value;
 }
 
-FArchive& FSpatialNetBitReader::operator<<(UObject*& Value)
```

Comment:
Why was this override removed?
I understand the behavior will remain correct, but keeping the override should avoid the extra call to `USpatialPackageMapClient::SerializeObject` from `FNetBitReader::operator<<(UObject*&)`

Reply:
Also, we still have the corresponding override in `FSpatialNetBitWriter`

Reply:
Interesting remark. Who choses how to write object references ? The Archive, or the PackageMapClient ?
I guess I'm not really consistent about that between the Reader and the Writer, but the architecture suggests it should be the PackageMapClient.
I will add back the override anyway.

Reply:
I would claim that if the `FSpatialNetBitReader` (or writer) is used with a package map different from `USpatialPackageMapClient`, something is going to break anyway. We can revisit this if at some point we want to have multiple spatial package maps or something, but for now this would just save an extra call and make the code a bit easier to follow in my mind.